design/open/hard-delete.md

```diff
@@ -0,0 +1,202 @@
+
+## 2. Reference Counter on Object Metadata
+
+This proposal suggested to maintain a reference counter in the underlying object's metadata and use the different block adapters
+capabilities to perform atomic, concurrently safe incrementation of this counter.
+The way we are handling the staged data in lakeFS, will not change in any other way.
+**The proposal is not viable since though Google Storage and Azure provide this capability, AWS S3 does not.** 
+
+This solution incurs an additional metadata read (and possible write) in case it is in the repo namespace in selected operations.
+
+### Blockers
+
+1. AWS does not support conditional set of metadata
+
+### Flows
+
+#### Upload Object
+
+1. Add a reference counter to the object's metadata and set it to one.
+2. Write the object to the object store the same way we do today.
+
+#### Stage Object
+
+1. Object path is in the repo namespace:
+   1. read its metadata, increment the counter and update the metadata on the object. 
+   2. If the object's write to object store fails - rollback the metadata update. 
+2. Object path is outside the repo namespace:
+   1. lakeFS will not handle its retention - therefore it will have to be deleted by the
+   user or by a dedicated GC process
+
+#### Delete Object
+
+1. Read object's metadata
+2. Reference counter exists:
+   1. if counter == 1
+      1. Hard-delete object
+   2. if counter > 1
+      1. Decrement counter 
+3. If reference counter doesn't exist:
+   1. Retention is not handled in lakeFS
+
+#### LinkPhysicalAddress
+
+Assume this API uses a physical address in the repo namespace
+1. Read object's metadata
+2. Increment counter and write metadata
+3. Add staging entry
+
+#### StageObject
+
+If physical address in repo namespace - perform increment as described above, otherwise do not handle retention
+
+#### Reset Branch
+
+When resetting a branch we throw all the uncommitted branch data.
+We can leverage the new KV drop async functionality to also check and hard-delete objects as needed
+
+### Atomic reference counter
+
+For each storage adapter (AWS, GC, AZ), we will use a builtin logic to provide this functionality
+
+#### AWS
+
+AWS doesn't allow to update an object or its metadata after creation. In order to update the metadata a copy object should be performed with the new metadata values. The only 
+conditionals currently available are on the objects timestamp and ETag. Unfortunately ETags are modified only on objects content change and does not take into account changes in metadata.
+
+#### Azure
+
+Store the reference counter in the blob's metadata and use `set blob metadata` API with the supported conditional header on the ETag to perform an If-Match operation and update
+the reference counter
+
+
+#### Google Store
+
+Store the reference counter in the blob's metadata, use the `patch object` API with the `ifMetagenerationMatch` conditional.
+
+
+## 3. Tracking references in staging manager
```

Comment:
Why is base assumption 1 valid here? I don't think we need to change the way we copy in this solution.

Reply:
Yes, deeper examining of the copy object flow show's this assumption is irrelevant for option 3